File: server/csharp/MetaObjects/MetaEndpoint.cs
using MetaObjects.Meta;

namespace MetaObjects;

/// <summary>
/// UI-1 — the metadata API contract.
///
/// C# ships the helper and not a mount: the runtime <c>MetaObjects</c> package
/// references only YamlDotNet, and an ASP.NET FrameworkReference here would put a
/// web dependency on every consumer of the core package. Mount it yourself:
///
/// <code>
/// app.MapGet($"/api{MetaEndpoint.MetaRoutePath}",
///            () => Results.Text(MetaEndpoint.MetaJson(root), "application/json"));
/// </code>
///
/// Guard that route with your own authorization — <c>/_meta</c> publishes the
/// shape of the model (entity and field names, types, validators, layouts),
/// though no row data.
/// </summary>
public static class MetaEndpoint
{
    /// <summary>The endpoint path, mounted under the host's API prefix. A cross-port
    /// contract value — one browser read-model works against every backend.</summary>
    public const string MetaRoutePath = "/_meta";

    /// <summary>The response body: the model as EFFECTIVE canonical JSON, so a browser
    /// reading it never has to resolve <c>extends</c> itself.</summary>
    public static string MetaJson(MetaData root) => SerializerJson.CanonicalSerializeEffective(root);
}

File: server/python/src/metaobjects/meta_endpoint.py
"""UI-1 — the metadata API contract.

Python ships the helper and not a mount: neither ``metaobjects.runtime`` nor
``metaobjects.codegen.runtime`` imports a web framework, and one endpoint is not
a reason to make either of them framework-bound. Mount it yourself::

    @app.get(f"/api{META_ROUTE_PATH}")
    def meta() -> Response:
        return Response(meta_json(root), media_type="application/json")

Guard that route with your own authorization — ``/_meta`` publishes the shape of
the model (entity and field names, types, validators, layouts), though no row data.
"""
from __future__ import annotations

from metaobjects.meta.meta_data import MetaData
from metaobjects.serializer_json import canonical_serialize_effective

#: The endpoint path, mounted under the host's API prefix. A cross-port contract
#: value — one browser read-model works against every backend.
META_ROUTE_PATH = "/_meta"


def meta_json(root: MetaData) -> str:
    """The response body: the model as EFFECTIVE canonical JSON.

    Effective, not raw: the effective form materializes the super-chain merge, so
    a browser reading it never resolves ``extends`` itself.
    """
    return canonical_serialize_effective(root)
